### PR TITLE
Issue/4728 show quick actions in notifications only if device is unlocked

### LINKED ISSUE #4728: Notifications: 2fa should only act on "Approve" if device screen is unlocked
### Expected behavior
When a 2fa notification is received, quick actions APPROVE and IGNORE are offered. If your screen is locked, APPROVE should tell you you can't approve while the screen is locked.

### Actual behavior
When you tap APPROVE, the approval is sent to wp.com even if the screen is locked.

### Steps to reproduce the behavior
- enable 2FA on your website
- log in to your app, then lock your phone's screen once logged in
- log in to your website on a desktop browser
- the 2FA auth push is received, with quick actions available and actionable.
- Tap "Approve" to log in to your site
- you get to log in even when the device screen is locked.

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4728 

This PR actually works for all notifications that have quick actions (that is, Comment notifications and the specific case of issue #4278 which is the 2FA push notification).
What it does is it re-creates the notification to only show quick actions if the device is unlocked and the app is unlocked as well.
The app now registers a broadcast receiver to check when the screen goes off, or when the screen is unlocked, and also checks the case that the user sets a PIN lock within WPAndroid app and still has a notification in the system dashboard.

To test:
CASE A: 2FA
- enable 2FA on your website
- log in to your app, then lock your phone's screen once logged in
- log in to your website on a desktop browser
- the 2FA auth push is received, with NO quick actions available.
- unlock the phone screen
- you should now see the quick actions APPROVE/IGNORE available in the push notification.
- If you lock the screen again, you should see the quick actions disappear from such notification (and viceversa, as many times as you do this).
- Tap "Approve" to log in to your site

CASE B: Comments
- log in to your app, then lock your phone's screen once logged in
- have some other account like a post of a blog you're an admin of
- the comment notification is received, with NO quick actions available.
- unlock the phone screen
- you should now see the quick actions REPLY/LIKE or REPLY/APPROVE (depending on your moderation settings for such blog) available.
- If you lock the screen again, you should see the quick actions disappear from such notification (and viceversa, as many times as you do this).

cc @hypest @kwonye @daniloercoli  
assigning to @hypest as he had the quick actions for 2FA notif idea originally :)

### CODE REVIEW

File: WordPress/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -1060,6 +1060,7 @@
     <string name="error_notif_q_action_like">Could not like comment. Please try again later.</string>
     <string name="error_notif_q_action_approve">Could not approve comment. Please try again later.</string>
     <string name="error_notif_q_action_reply">Could not reply to comment. Please try again later.</string>
+    <string name="error_notif_2fa_device_locked">Please unlock your device before performing this action</string>
```
Comment (daniloercoli) [inline WordPress/src/main/res/values/strings.xml]: There is a missing period at the end of the sentence.

**Comment by hypest** [comment]
Thanks for the new PR @mzorz!

I've been trying out the "A" case (the 2FA notification) but couldn't make it work reliably. I notice in partivular that the mutation doesn't seem to work if the app is not on the foreground. For example, try this:
* Go to the home screen and keep it turned ON
* Try to log in to your 2FA site
* A notification will arrive. Slide it down and notice the quick actions being there.
* Turn the screen OFF (and assuming you have screen lock enabled...)
* Turn it ON again and slide the notification... the quick actions will still be there. At this point I was expecting the actions to have been removed. Have I understood the feature incorrectly maybe?

Let me know If am Holding It Wrong™️ :)

**Comment by mzorz** [comment]
`<facepalm/>` , you're actually holding it right! I missed testing a case after a quick refactor, let me check it all again, commit what I missed and get back to you with it. Apologies for the time taken!

**Comment by mzorz** [comment]
fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/4736/commits/4a5b573e5ff31d8069943f58ad8b529b6535a468  - ready for another pass 🙇

**Comment by hypest** [comment]
It seems to still not work on my end. I compiled and ran the app again, went to home screen and then did the login routine but the quick actions are still there on the lockscreen. Here's a screencast: https://cloudup.com/iZJp1OcMjq1

**Comment by mzorz** [comment]
Ok after having discussed on Slack, @hypest detected that when the app is not in the foreground the behavior was still the same (i.e. quick actions still are there on lock screen), so made a sticky service that registers the broadcast receiver to make sure we are notified of the screen_off and user_present events and this way are able to add/remove quick actions accordingly. 
Tested on Nexus 5 with Android 6.0.1, and emulator with Android N

**Comment by hypest** [comment]
There seems to be an issue now where the auth notification gets duplicated when the screen is turned off and the lock screen now shows 2 notifications.

**Comment by mzorz** [comment]
Thank you @hypest! fixed the dupe case here https://github.com/wordpress-mobile/WordPress-Android/pull/4736/commits/a30872377cf781a6feb9a3841ccfec4120da43d7

cc @daniloercoli would you like to give it a try? reassigning with permission of @hypest

**Comment by daniloercoli** [comment]
Works on Nexus 5X with Android 7.X. 
We should just disable the notification sound when recreating the notifications.

**Comment by mzorz** [comment]
Fixed the sounds handling in https://github.com/wordpress-mobile/WordPress-Android/pull/4736/commits/b09b1faf61958eb811e6b57cfa8ea0249736dfbc

**Comment by daniloercoli** [comment]
:shipit:

**Pull request merged**